whitepapers/0013_ccq.md

```diff
@@ -0,0 +1,350 @@
+
+For these and other reasons, a separate REST server is being provided to accept off-chain query requests, submit them to the gossip network, correlate the responses, and return them to the caller.
+
+The CCQ server will be configured to publish and listen on the CCQ P2P network. It's P2P peer ID will be registered with the guardians. The server will join both `ccq_req` and `ccq_resp` but only subscribe to `ccq_resp`.
+
+The server will be configured with a list of supported users. Each user will have an API Key, as well as a list of what queries they are allowed to perform.
+The query data per user will look something like this:
+
+```
+{
+    "userName": "Test User",
+    "apiKey": "432185ac-fbb7-4883-b68d-888e7bd749a9",
+    "allowUnsigned": true,
+    "allowedCalls": [
+    {
+        "ethCall": {
+        "note:": "Name of WETH on Goerli",
+        "chain": 2,
+        "contractAddress": "B4FBF271143F4FBf7B91A5ded31805e42b2208d6",
+        "call": "0x06fdde03"
+        }
+    }
+    ]
+}
+```
+
+When a user submits a query, they will need to set the `X-API-Key` header in the request.
+
+The server will only accept requests from configured users. Those requests may only be for the RPC calls configured in the `allowedCalls` list.
+
+Each allowed call is specified by:
+
+- Target wormhole chain ID
+- Target contract address on that chain
+- The first four bytes of the hash of signature of the method to be called.
+
+All configured users may submit queries that they sign with their own key. In addition to signed requests, if the `allowUnsigned` flag is set to `true`, the user may submit unsigned requests and the server will sign them using a pre-configured key. Note that all keys must be in the guardian allow list.
+
+## Typescript Library
+
+To assist integrators, there is a typescript library in `sdk/js-query` that provides methods for building queries. Additionally, there are some usage examples in `sdk/js-query/src/query/ethCall.test.ts`.
+
+## EVM Response Parsing Library
+
+Once a response has been received, the integrator needs to validate and parse it to get the results. There is a simple EVM library to facilitate this which can be found in
+`ethereum/contracts/query/QueryResponse.sol`. A usage example is in `ethereum/forge-test/query/Query.t.sol`.
+
+## Core Bridge Optimization
+
+The `QueryResponse` parsing library uses the `Wormhole Core` contract to verify the signatures on the response. This currently involves making four cross contract calls, wasting gas.
+A necessary optimization is to add a new public method to the `Wormhole` contract that will take a payload (the response bytes) and the set of signatures, and do the verification in a single call. It should revert if the signature is not valid.
```

Comment:
I'd change this language to be more "maybe" vs "will".

e.g. `This currently involves making four cross-contract calls, each spending gas. A possible optimization is to...`